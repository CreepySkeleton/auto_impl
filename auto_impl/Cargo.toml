[package]
name = "auto_impl"
version = "0.1.0"
authors = ["Ashley Mannix <ashleymannix@live.com.au>"]

[lib]
proc-macro = true

[dependencies]
quote = "*"
auto_impl_internals = { version = "*", path = "../auto_impl_internals" }