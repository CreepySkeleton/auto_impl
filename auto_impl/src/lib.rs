#![feature(proc_macro)]

extern crate quote;
extern crate proc_macro;
extern crate auto_impl_internals;

use std::str::FromStr;
use quote::Tokens;
use proc_macro::TokenStream;

fn auto_impl_expand(attrs: TokenStream, proc_tokens: TokenStream) -> Result<TokenStream, String> {
    let mut tokens = Tokens::new();
    tokens.append(proc_tokens.to_string());

    let mut attr_tokens = Tokens::new();
    attr_tokens.append(format!("#[auto_impl{}]", attrs.to_string()));

    let impl_types = derive_internals::parse_impl_types(attr_tokens)?;

    let tokens = derive_internals::auto_impl_expand(&impl_types, tokens)?;

    TokenStream::from_str(&tokens.to_string()).map_err(|e| format!("{:?}", e))
}

#[proc_macro_attribute]
pub fn auto_impl(attrs: TokenStream, proc_tokens: TokenStream) -> TokenStream {
    match auto_impl_expand(attrs, proc_tokens) {
        Ok(tokens) => tokens,
        Err(e) => panic!("auto impl error: {}", e)
    }
}
