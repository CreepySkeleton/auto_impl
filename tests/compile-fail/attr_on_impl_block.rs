#![feature(use_extern_macros)]

extern crate auto_impl;
use auto_impl::auto_impl;


trait Foo {}

#[auto_impl(&, &mut)]
impl Foo for usize {}
