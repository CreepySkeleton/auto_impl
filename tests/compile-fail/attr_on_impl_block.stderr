error: couldn't parse trait item

note: expected `trait`

note: the #[auto_impl] attribute can only be applied to traits!
 --> $DIR/attr_on_impl_block.rs:7:1
  |
7 | impl Foo for usize {}
  | ^^^^
